### PR TITLE
In-form caching

### PR DESCRIPTION
**Pull request by amstone326**
cross-request: https://github.com/dimagi/commcare-core/pull/716

### CODE REVIEW

**Comment by amstone326** [comment]
@damagatchi retest this please

**Comment by amstone326** [comment]
@damagatchi retest this please

File: app/unit-tests/src/org/commcare/android/tests/queries/ExpressionCacheTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        TestUtils.initializeStaticTestStorage();
+        TestUtils.processResourceTransaction("/inputs/case_test_model_query_lookups.xml");
+    }
+
+    @Test
+    public void testFasterWithCaching() {
+        long timeWithoutCaching = time100Evaluations(false);
+        long timeWithCaching = time100Evaluations(true);
+        Assert.assertTrue("Evaluation time with caching should be shorter", timeWithCaching < timeWithoutCaching);
```
Comment (ctsims) [inline app/unit-tests/src/org/commcare/android/tests/queries/ExpressionCacheTest.java]: Hm, if this difference is huge, I think that makes sense. I might push for simply counting the cache hit / miss ratio or something similar? Timing is such an awkward thing to test in some ways.

**Comment by amstone326** [comment]
@damagatchi retest this please

File: app/unit-tests/src/org/commcare/android/tests/queries/ExpressionCacheTest.java (hunk 1 above)
Comment (amstone326) [reply app/unit-tests/src/org/commcare/android/tests/queries/ExpressionCacheTest.java]: @ctsims I added a new test below for testing the number of expression evaluations that used the cache, but the time differential here is actually very substantial. `timeWithoutCaching` is consistently almost twice as large as `timeWithCaching`. Given that, are you ok with leaving it?

**Comment by ctsims** [comment]
@amstone326 Approving these, but they both have "don't pull" set. Wasn't sure if you wanted anything else done. Let me know if you want me to straight-pull instead.

**Comment by amstone326** [comment]
@damagatchi retest this please

**Comment by amstone326** [comment]
@ctsims Sorry about that, wasn't intentional!

**Pull request merged**